File: app/src/main/java/com/halfplatepoha/jisho/JishoMigration.java
package com.halfplatepoha.jisho;

import com.halfplatepoha.jisho.jdb.Schema;

import javax.inject.Inject;

import io.realm.DynamicRealm;
import io.realm.FieldAttribute;
import io.realm.RealmMigration;
import io.realm.RealmObjectSchema;
import io.realm.RealmSchema;

/**
 * Created by surjo on 13/05/17.
 */

public class JishoMigration implements RealmMigration {

    @Inject
    public JishoMigration() {}

    @Override
    public void migrate(DynamicRealm realm, long oldVersion, long newVersion) {
        RealmSchema schema = realm.getSchema();

        if(oldVersion <= 0) {
            RealmObjectSchema sense = schema.get("FavSense");
            if(sense.hasField("verbType"))
                sense.removeField("verbType");
            oldVersion++;
        }

        if(oldVersion < Jisho.APP_REALM_VERSION) {
            
            schema.get(Schema.Entry.TABLE_NAME)
                    .addField(Schema.Entry.NOTE, String.class);

            schema.create(Schema.JishoList.TABLE_NAME)
                    .addField(Schema.JishoList.NAME, String.class, FieldAttribute.REQUIRED)
                    .addRealmListField(Schema.JishoList.ENTRIES, schema.get(Schema.Entry.TABLE_NAME));

        }
    }
}

File: app/src/main/java/com/halfplatepoha/jisho/v2/realm/JishoMigration.kt
package com.halfplatepoha.jisho.v2.realm

import io.realm.DynamicRealm
import io.realm.RealmMigration

/**
 * Created by surjo on 13/05/17.
 */

class JishoMigration : RealmMigration {

    companion object {
        const val DB_VERSION = 3L
    }

    override fun migrate(realm: DynamicRealm, oldVersion: Long, newVersion: Long) {
        val schema = realm.schema

        if (oldVersion == 0L) {
            val sense = schema.get("FavSense")
            if (sense!!.hasField("verbType"))
                sense.removeField("verbType")
        }

        if(oldVersion < 3L) {
            schema.create("HistoryV2")
                    .addField("id", String::class.java)
                    .addField("timestamp", Long::class.java).setNullable("timestamp", true)
                    .addField("word", String::class.java)
            schema.create("FavoriteV2")
                    .addField("id", String::class.java)
                    .addField("word", String::class.java)
                    .addField("furigana", String::class.java)
        }
    }

    override fun equals(other: Any?): Boolean {
        return other != null && other is JishoMigration
    }

    override fun hashCode(): Int {
        return 122
    }
}
